src/bp/ui-admin/src/telemetry/client.ts

```diff
@@ -0,0 +1,146 @@
+  }
+  return 0
+}
+
+const addEventTimeout = (event: string, timeout: number) => {
+  setTimeout(() => toggleLock(event), timeout)
+  window.localStorage.setItem(event, (timeout + moment().valueOf()).toString())
+}
+
+const checkStoreInfoReceived = () => !Object.keys(storeInfos).find(info => _.isEmpty(storeInfos[info]))
+
+export const addStoreInfo = (name: string, pathInStore: string, formatter?: Function) => {
+  storeInfos[name] = {
+    storedInfo: '',
+    loadInfo: () => {
+      const value = _.get(store.getState(), pathInStore)
+      storeInfos[name].storedInfo = formatter ? formatter(value) : value
+    }
+  }
+}
+
+export const getStoreInfo = (name: string) => {
+  if (storeInfos[name]) {
+    return storeInfos[name].storedInfo
+  } else {
+    throw `The information "${name}" asked is not tracked by the telemetry module. Maybe it was not added before it's use?`
+  }
+}
+
+export const addTelemetryEvent = (name: string, timeout: string, getPackage: Function) => {
+  eventPackageInfo[name] = {
+    locked: getEventTimeout(name) >= 0,
+    timeout,
+    getPackage
+  }
+}
+
+const checkTelemetry = async () => {
+  for (const eventName in eventPackageInfo) {
+    if (!eventPackageInfo[eventName].locked) {
+      toggleLock(eventName)
+
+      await sendTelemetry(getTelemetryPackage(eventName, eventPackageInfo[eventName].getPackage()), eventName)
+    }
+  }
+}
+
+export const startTelemetry = () => {
+  addStoreInfo('email', 'user.profile.email')
+
+  addStoreInfo('bp_release', 'version.currentVersion')
+
+  addStoreInfo('bp_license', 'license.licensing.isPro', (info: any) => {
+    return !!info ? 'pro' : 'community'
+  })
+
+  addTelemetryEvent('ui_language', '8h', () => {
+    return {
+      user: {
+        email: getStoreInfo('email'),
+        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone
+      },
+      language: lang.getLocale()
+    }
+  })
+
+  for (const event in eventPackageInfo) {
+    if (getEventTimeout(event) >= 0) {
+      setTimeout(() => toggleLock(event), getEventTimeout(event))
+    }
+  }
+
+  store.subscribe(() => {
+    for (const infoName in storeInfos) {
+      storeInfos[infoName].loadInfo()
+    }
+  })
+
+  setInterval(() => {
+    if (checkStoreInfoReceived()) {
+      checkTelemetry().catch(err => {
+        console.error(err)
+      })
+    }
+  }, ms('30s'))
+}
+
+const getTelemetryPackage = (event_type: string, data: object): TelemetryPackage => {
+  return {
+    schema: telemetryPackageVersion,
+    uuid: uuid.v4(),
+    timestamp: new Date().toISOString(),
+    bp_release: getStoreInfo('bp_release'),
+    bp_license: getStoreInfo('bp_license'),
+    event_type: event_type,
+    source: 'client',
+    event_data: {
+      schema: dataClusterVersion,
+      ...data
+    }
+  }
+}
+
+const sendTelemetry = async (data: TelemetryPackage, event: string) => {
+  await axios.post('/', data, axiosConfig)
+  addEventTimeout(event, ms(eventPackageInfo[event].timeout))
```

Comment:
what's that timeout ? why is this needed ? I feel like there's way too much unnecessary logic

 if its sent then it's sent and just post a success to bp server

Reply:
The timeout is to store the moment when the info can be sent again. If I take the example of the "ui_language" event, I configured this package to be sent only every 8h. This line make sure that the package can be sent again in 8 hours whether the browser is closed or not